auv_sim/auv_sim_bridge/include/auv_sim_bridge/simulation_mock_ros.h

```diff
@@ -81,24 +120,32 @@ class SimulationMockROS {
     depth_pub_.publish(depth_msg);
   }
 
-  void dvlCallback(const uuv_sensor_ros_plugins_msgs::DVL &msg) {
+  void altimeterCallback(const sensor_msgs::Range &msg) {
     std_msgs::Float32 altitude_msg;
-    geometry_msgs::Twist velocity_msg;
-    geometry_msgs::TwistWithCovarianceStamped velocity_stamped_msg;
-    // Dvl altitude
-    altitude_msg.data = msg.altitude;
+
+    altitude_msg.data = msg.range;
     altitude_pub_.publish(altitude_msg);
 
+    latest_altitude_ = msg.range;
+  }
+
+  void dvlCallback(const nav_msgs::Odometry &msg) {
+    geometry_msgs::Twist velocity_raw_msg;
+    std_msgs::Bool is_valid_msg;
+
+    if (latest_altitude_ > 0.3 && dvl_enabled_) {
+      velocity_raw_msg = addNoiseToTwist(msg.twist.twist);
+      velocity_raw_msg = rotateVelocity(velocity_raw_msg, 135.0);
+      is_valid_msg.data = true;
+    }
+
     // Dvl velocity
-    velocity_msg.linear = msg.velocity;
-    velocity_pub_.publish(velocity_msg);
-
-    boost::array<double, 36> covariance;
-    covariance.fill(1e-6);
-    velocity_stamped_msg.header.frame_id = "dvl_link";
-    velocity_stamped_msg.twist.twist = velocity_msg;
-    velocity_stamped_msg.twist.covariance = covariance;
-    velocity_stamped_pub_.publish(velocity_stamped_msg);
+    velocity_raw_pub_.publish(velocity_raw_msg);
+
+    // Dvl validity
+    is_valid_pub_.publish(is_valid_msg);
+
+    dvl_rate_.sleep();
```

Comment:
Sleeping here would create queue problems, instead you should store the latest reading, and run your modifications on a different loop which rolls at dvl rate. Use ROS timer perhaps to detach a different process which dies not block the loop

Reply:
I understand why this would create a problem. I tried using a ROS timer to create a different callback with dvl rate, but it seems to have a problem. This is a log of the implementation in the current code, I will investigate this further.

```
[1733629144.805371388, 2.100000000]: TIMER CALLBACK
[1733629145.272659647, 2.250000000]: TIMER CALLBACK
[1733629145.428972508, 2.300000000]: TIMER CALLBACK
[1733629145.741480176, 2.400000000]: TIMER CALLBACK
[1733629146.052502991, 2.500000000]: TIMER CALLBACK
[1733629146.361058481, 2.600000000]: TIMER CALLBACK
[1733629146.673469617, 2.700000000]: TIMER CALLBACK
[1733629147.133962142, 2.850000000]: TIMER CALLBACK
[1733629147.445909725, 2.950000000]: TIMER CALLBACK
[1733629147.604225889, 3.000000000]: TIMER CALLBACK
```

Reply:
Hmm, yeah that seems to be a bit odd. Maybe it is related to the RTF if your node is not using simulation time. Or even taking it back, maybe the timestamps that you see is the real time instead of the sim time, thus there is a difference. Anyways I dint want to point you into the wrong direction so feel free to debug this any way you like